### DISCUSSION TITLE
Introduction

### DISCUSSION THREAD

**Discussion by WebMechanic** [Stegosource/vuetensils#148] [General]
To keep this out of [this issue....](https://github.com/AustinGil/vuetensils/pull/130)

Hello,
I'm in Germany working on a big relaunch of our real estate portal.
We're still evaluating Vuetensils, only screened a couple so far. Markup and styles are sometimes difficult to examine to see if and how we can apply our stuff. We basically slap our own class lists on them and try to keep our hand off your namespace.

We have a minimalistic approach to markup and work a lot with CSS Grid to avoid that pesky nested div soup. Inline grids are great for micro layouts of form controls and widgets. We almost exclusively nest all controls inside their label element (few exceptions) then use the label as The Container™ for white space and positioning.
Hints and messages are rendered in (optional) sibling elements of the control so we can leverage many CSS state selectors and avoid JavaScript just for styling.
That's one of the reasons I don't think the form controls will make it into our project, but I got some inspirations 😇

**Comment by AustinGil** [comment]
Right on. Yeah, I actually want to move labels to be siblings of inputs and not wrap them. I think there are flaws with wrapping labels that I was not aware of before.

Regarding the namespace thing, using your own classes is smart to avoid any future issues, but yeah, the classes I use are intended for others to tap into for styling. Many elements have classes without any associated styles. Maybe this adds unnecessary characters. I could consider removing them and ONLY using classes I need styles for and expect folks to inject their own classes.

CSS :has should be a big win for styling without JS. I've got a video for this coming up soon :)

The form stuff is one of my favorite parts of this project. Still needs some work, but you should at least consider forking it for your needs. I make all the components RELATIVELY easy to copy and make your own from the source code

**Reply by WebMechanic** [reply]
> CSS :has should be a big win for styling without JS. I've got a video for this coming up soon :)

nice, I'll watch :D

Unfortunately we still have many old iOS Safaris so `:has` has to wait as a general tool and I don't wanna polyfill it, eso. not on old phones.
We do use it laptop/desktop and up layouts.

> The form stuff is one of my favorite parts of this project.

one can tell actually, I like how you implement them. However, the markup just doesn't suite us and validation is also handled differently.
We probably use them in our new CRM areas, where there are many more forms and fewer layout and design constraints.

**Reply by AustinGil** [reply]
Yeah, I actually have an issue in Vuetensils because I'm using :where, but I need to roll that change back for the time being.

**Reply by WebMechanic** [reply]
yep, seen it :) 
I simply created a separate .scss file in our project and included the `main.css` from your src/ folder. Did that for other "vendor" styles, too.
That way I can forward it for page styles that need it and minification for production is done anyway.

**Comment by WebMechanic** [comment]
> I think there are flaws with wrapping labels that I was not aware of before.

care to elaborate?

I find a nested setup easier to manage and style, and you get a larger hit area for free, useful on mobile (we have +65% mobile users), **no need for an id/for** which I hated since ever. 

On the JS side you can easily travel up/down this small subtree, incl. the standard references between the label and its control and things like `.that-form [name="fubar"]` make strong and pretty unique selectors for both CSS and JS.

The general layout for most of our form fields follows this raw pattern:
```html
<label><b>The usual bold Text</b>
  <span class="hint">Pointless explanation of what characters are required for
  a <q>safe</q> password, although a <em>passphrase</em> would have higher 
  entropy and be much safer and much easier to remember.</span>
  <input type="password" />
  <span class="error">A password only works if you enter some characters. Just sayin'.</span>
</label>
```
The pack is great to style imho. Make `b` and `span` being `display: block` and you have a nice stack; dito with even less CSS if you make the label itself a `grid`. And :focus / :invalid / :valid / :placeholder-shown all give nice selectors for the input and the error message.

**Reply by AustinGil** [reply]
Yeah. I think the larger hit area is generally better, but it can be weird in some scenarios. Say you wanted to have a list of radio inputs that are vertically stacked. You might make them full width, or use a 1 column grid, or a flex container with the column direction. Visually, the elements would look fine, but if the label wraps the input, the hit box can be SO large that some white space all the way across the screen is part of it. This is weird to me. So I prefer having an inline-block div wrap each form control that can take that responsibility.

It's still good to have the label increase the hit area for an input, to be sure. But you get that for free if you use the `for` attribute with the input's ID anyway. And I see that you don't like that approach, neither do I, but it's actually slightly better for accessibility as the wrapping method (implicit) doesn't work as well as the ID method (explicit) when dealing with voice control technology (specifically Dragon Speech Recognition). So even if/when I use a wrapping label, I still put the ID anyway.

**Reply by WebMechanic** [reply]
I see, thanks.
Voice control is sth we sure need to evaluate - thanks for the reminder :) 

I recall using Nuance years ago and it sucked with non english language accents, so I don't know if they have that much of a relevant market share where we are. 
Stuff will always break for somebody...

**Comment by WebMechanic** [comment]
> Many elements have classes without any associated styles. Maybe this adds unnecessary characters. 

yeah, but those extra characters do no harm. It's not like you're polluting the `class` attributes in a Tailwind manner ;-)

> I could consider removing them and ONLY using classes I need styles for and expect folks to inject their own classes.

nah, keep 'em. 
It's still interesting to see what states and features you had in mind, when you added them. 
If there's i.e. some feature we haven't considered or just forgot, we actually use your class names instead of inventing the same thing. Given the vts- namespace, it's ease to pin the target.

We virtually wrote everything[^1] from scratch and decided against using any 3rd party design || layout system and their fake one dimensional "grid systems" based on Flexbox. This stuff always needs a butt-load of extra markup just for these "columns" to work. Not adding this dependency also keeps our style sheets free from namespaces that aren't ours.  
We also have no use for 20, mostly inaccessiblee, color ramps where no base colour is even close to our 3 main CD colour shades.

We also ditched many "big" (Vue) UI packs that would have forced us too, to submit to both their idea of organising and naming even more styles we don't need nor care for, eventually adding bloat overwriting it, and adopt a (non-semantic) markup structure for that whole bloody thing to work.

Your components being "naked" _and_ accessible was the **primary selling point**, so: many thanks for going that road 👍🏻♥

[^1]: we kickstarted with Pico-CSS, shifted things around, then dropped or modified what didn't match our concepts, and included many of Open Props.

**Reply by AustinGil** [reply]
That's cool. You are basically describing the ideal user for Vuetensils. I don't think it will ever be one of the major Vue libraries, but it's in a good place right now for what it is and mostly does what it needs to pretty well :)
